### PR TITLE
Allow to ask students to place points and lines.

### PR DESCRIPTION
**Pull request by smarnach**
This PR add support for lines and points to be placed by students as part of the answer.

The support for lines was based on an [implementation by @ChristopherChudzicki](https://github.com/ChristopherChudzicki/jsinput-vectordraw/commits/type-equal-line).

For testing, modify the file `api-example.html` and add objects of the new types, e.g. add something like

    {name: 'a', description: 'Line - a', type: 'line'}

to the `vectors` array, or something like

    {name: 'p', description: 'Point - p', coords: [1, 1], fixed: false, render: false}

to the `points` array.  Also make sure to test some checks for the new object types.

The UI for points needs some careful testing.  They are a fundamentally new object type, and there are many corner cases when placing and dragging points and vectors, in particular when placing them on top of each other.  I'm currently not aware of any remaining issues, but quite a few came up during local testing.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -102,9 +106,23 @@ These are the supported vector properties:
...
+The `points` setting defines a list of points to be either just drawn
+on the board for reference, or to be placed by the student.  Points can
+have the following properties:
+
+* `name` (required): The unique name of the point, used in checks and
+  internally to identify the point.
+* `coords` (required): An array of length 2 with the (initial)
+  coordinates of the point.
+* `fixed`: Whether the point should be drawn in a fixed location or
+  placed by the student.  Defaults to `true`.
```
Comment (e-kolpakov) [inline README.md]: @smarnach ... and reading this I have no idea what `true` stands for. :) Well, actually, since I'm familiar with how this thing works internally I do, but someone else will probably be confused by this.

File: README.md [hunk 2]
```diff
@@ -102,9 +106,23 @@ These are the supported vector properties:
...
+* `render`: Whether the point should be rendered when the board is
```
Comment (e-kolpakov) [inline README.md]: @smarnach "render" is more of a technical term, while this readme is mostly user-oriented. Consider changing to "shown"

File: test_vectordraw.py [hunk 3]
```diff
@@ -183,6 +183,22 @@ def test_check_segment_angle(self):
         self.assertFails(vd.check_segment_angle(self.check(-45, 9, errmsg=custom_errmsg), vectors),
                          custom_errmsg)
 
+    def test_check_points_on_line(self):
+        errmsg = 'The line vec does not pass through the correct points.'
+        vectors = {'vec': self.vector(1, 1, 5, 5)}
+        self.assertPasses(vd.check_points_on_line(self.check([[1, 1], [5, 5]]), vectors))
+        self.assertPasses(vd.check_points_on_line(self.check([[1, 2], [5, 4]]), vectors))
+        self.assertPasses(vd.check_points_on_line(self.check([[3.7, 2.3], [2.3, 3.7]]), vectors))
```
Comment (e-kolpakov) [inline test_vectordraw.py]: @smarnach nit: I would argue that (1,2) lies on (1,1) -> (5,5) line. Default tolerance is way too big here. So I would suggest passing lower tolerance to `self.check`. Feel free to ignore it, but I understood that tolerance in play only by looking at the rest of the code.

File: vectordraw.js [hunk 4]
```diff
@@ -135,6 +135,48 @@ VectorDraw.prototype.createBoard = function() {
...
+    var opts = {
+        size: 1,
+        fixed: point.fixed,
+        name: point.name,
+        withLabel: false,
+        strokeColor: 'pink',
+        fillColor: 'pink',
+        showInfoBox: false
+    };
+    if (opts.fixed === undefined) {
```
Comment (e-kolpakov) [inline vectordraw.js]: @smarnach global nit: I don't like that there're a lot of `something === undefined` checks. Maybe it would make sense to add a more intelligent "parse settings" step to initialization code instead of a `this.settings = _.extend(default_settings, settings);`?

File: vectordraw.js [hunk 5]
```diff
@@ -210,7 +252,7 @@ VectorDraw.prototype.renderVector = function(idx, coords) {
     tip.label.setAttribute({fontsize: 18, highlightStrokeColor: 'black'});
 
     // Disable the <option> element corresponding to vector.
-    var option = this.element.find('.menu option[value=' + idx + ']');
+    var option = this.element.find('.menu option[value=vector-' + idx + ']');
```
Comment (e-kolpakov) [inline vectordraw.js]: @smarnach nit: might make sense to extract method `getMenuOption(option_value)`

File: vectordraw.js [hunk 6]
```diff
@@ -222,16 +264,31 @@ VectorDraw.prototype.removeVector = function(idx) {
...
+        var option = this.element.find('.menu option[value=vector-' + idx + ']');
         option.prop('disabled', false);
     }
 };
 
-VectorDraw.prototype.addVectorFromList = function() {
+VectorDraw.prototype.getSelectedElement = function() {
+    var selector = this.element.find('.menu select').val();
+    if (selector) {
+        selector = selector.split('-');
```
Comment (e-kolpakov) [inline vectordraw.js]: @smarnach nit: looks a bit like a string typization. I believe using data attributes would make it cleaner.

File: vectordraw.js [hunk 7]
```diff
@@ -305,17 +362,24 @@ VectorDraw.prototype.updateVectorProperties = function(vector) {
...
     $('.vector-prop-name .value', this.element).html(vector.point2.name); // labels are stored as point2 names
-    $('.vector-prop-length .value', this.element).html(length.toFixed(2) + ' ' + length_units);
-    $('.vector-prop-angle .value', this.element).html(angle.toFixed(2));
+    if (vector.elType !== "line") {
+        $('.vector-prop-length .value', this.element).html(length.toFixed(2) + ' ' + length_units);
+        $('.vector-prop-angle .value', this.element).html(angle.toFixed(2));
+    }
+    else {
+        $('.vector-prop-length', this.element).hide();
+        $('.vector-prop-angle', this.element).hide();
```
Comment (e-kolpakov) [inline vectordraw.js]: @smarnach lines have angle too. Also, it ignores `show_slope_for_lines` setting - either restore it or remove from documentation

File: vectordraw.js [hunk 8]
```diff
@@ -329,7 +393,7 @@ VectorDraw.prototype.canCreateVectorOnTopOf = function(el) {
     if (el instanceof JXG.Point) {
         var vector = this.getVectorForObject(el);
         if (!vector) {
-            return true;
+            return !(el instanceof JXG.Point) || el.getProperty('fixed');
```
Comment (e-kolpakov) [inline vectordraw.js]: @smarnach we're in `el instanceof JXG.Point` branch, so first part of condition is always false

File: vectordraw.js (hunk 7 above)
Comment (e-kolpakov) [inline vectordraw.js]: @smarnach actually, there's a bug: if you select a line it hides the angle and length properties, but when you select a vector again those props are not shown back. Possible fixes:

1. Do not hide length and angle; for line use ∞ for length, show actual value for angle
2. In the "True" clause first show the hidden elements

File: README.md [hunk 9]
```diff
@@ -102,9 +106,23 @@ These are the supported vector properties:
...
+  initially shown.  Defaults to `true`, but may be set to `false` for
+  non-fixed points.
+* `description`: The long description used in the drop-down menu for
+  non-fixed points.
+* `style`: An object with addional properties for the point.  Supports
+  the properties `size`, `strokeColor` and `fillColor`.
```
Comment (e-kolpakov) [inline README.md]: @smarnach nit: supporting both `strokeColor` and `fillColor` for "Point" seems a bit counter-intuitive for me. It might be more intuitive to have a `color` property that would expand into both `strokeColor` and `fillColor`.

**Comment by e-kolpakov** [comment]
@smarnach also, the was a `slope` parameter shown for the line, but looks like you accidentally removed displaying it - please restore.

File: README.md (hunk 1 above)
Comment (smarnach) [inline README.md]: Clarified.

File: README.md (hunk 2 above)
Comment (smarnach) [inline README.md]: I used "displayed" since there is another use of the word shown later in the same sentence.

File: test_vectordraw.py (hunk 3 above)
Comment (smarnach) [inline test_vectordraw.py]: I used the same default tolerance as everything else (1.0 for distances, 2.0 for angles).  Since these are arbitrary units, it's completely unclear whether a distance of 1 is big or small.  I think it's OK fot the test – it should just verify it does what it is supposed to do.

File: vectordraw.js (hunk 4 above)
Comment (smarnach) [inline vectordraw.js]: I've implemented this suggestion, and I think it indeed improves the readability of the code.  Thanks!

File: vectordraw.js (hunk 5 above)
Comment (smarnach) [inline vectordraw.js]: Done.

File: vectordraw.js (hunk 6 above)
Comment (smarnach) [inline vectordraw.js]: @e-kolpakov You have a point here.  When we change this, we would have to stop using the `value` attribute of `option`, since it can only take string values, and that would make the code harder to read and more involved in other regards.  Now that I have introduced the `getMenuOption()` function, dealing with the strings is isolated in just two small functions, so I feel it's acceptable now.

File: vectordraw.js (hunk 7 above)
Comment (smarnach) [inline vectordraw.js]: I went for a mixture of those two – I now hide the length for lines, and show it again for other types.

File: vectordraw.js (hunk 8 above)
Comment (smarnach) [inline vectordraw.js]: I removed the first part of the confition.

File: README.md (hunk 9 above)
Comment (smarnach) [inline README.md]: Good suggestion, done.

**Comment by smarnach** [comment]
@e-kolpakov I've added back the slope display (and fixed the bugs it had).  I think I addressed all review comments, so please take another look.

**Comment by antoviaque** [comment]
@dseaton FYI - this is the improvement for vector drawing we discussed, if you want to review before this gets merged.

**Comment by e-kolpakov** [comment]
@smarnach you've forgot to remove unit_vector_ratio from show_slope_for_lines description

File: vectordraw.js [hunk 10]
```diff
@@ -1,6 +1,25 @@
 'use strict';
 
 var VectorDraw = function(element_id, settings) {
+    this.board = null;
+    this.dragged_vector = null;
+    this.drawMode = false;
+    this.history_stack = {undo: [], redo: []};
+    this.settings = this.sanitiseSettings(settings);
```
Comment (e-kolpakov) [inline vectordraw.js]: @smarnach nit: not technically a typo, but other parts of codebase use "American" spelling of -ze, not -se.

File: vectordraw.js [hunk 11]
```diff
@@ -297,25 +394,34 @@ VectorDraw.prototype.getVectorSettingsByName = function(name) {
...
-    $('.vector-prop-length .value', this.element).html(length.toFixed(2) + ' ' + length_units);
-    $('.vector-prop-angle .value', this.element).html(angle.toFixed(2));
+    if (vector.elType !== "line") {
+        $('.vector-prop-length', this.element).show();
+        $('.vector-prop-slope', this.element).hide();
+        $('.vector-prop-length .value', this.element).html(length.toFixed(2) + ' ' + vec_settings.length_units);
+        $('.vector-prop-angle .value', this.element).html(angle.toFixed(2));
+    }
+    else {
+        $('.vector-prop-length', this.element).hide();
```
Comment (e-kolpakov) [inline vectordraw.js]: @smarnach angle is shown for lines, but not updated - so it's frozen to the last displayed value

**Comment by e-kolpakov** [comment]
@smarnach :+1: conditional the angle is updated for lines and `unit_vector_ratio` mention is removed from readme

File: vectordraw.js (hunk 10 above)
Comment (smarnach) [inline vectordraw.js]: Fixed.

File: vectordraw.js (hunk 11 above)
Comment (smarnach) [inline vectordraw.js]: Fixed.

**Comment by smarnach** [comment]
@e-kolpakov Thanks again – I've applied all changes you suggested.  I'll add an attribution for the line support code somewhere, and wait for @dseaton to take a look before merging.

**Comment by dseaton** [comment]
@smarnach @antoviaque @ChristopherChudzicki 

Here are some comments:

1. What purpose does fixed serve for points? If it has purpose, why do vectors not have a fixed parameter?

2. For the properties box, do the values (slope, length, etc) update when a user selects an object? Say I had a problem with vectors and lines - would the display toggle?

3. I don't have a good suggestion at the moment, but is it possible to have a single set of variables that define grading parameters for points, lines, and vectors? For example, coord_1 and coord_2 could be used for any object, but it could be documented - coord 1 is all that matter for points, coord 1 and 2 matter for lines but not their order, and for vectors, the order matters. Don't set off to do this without discussing first. This may make it less clear to new users. If we choose to keep things in their current state, I would lobby that the documentation be updated such that a clear distinction is made between grading each type of object. **Note, the single set of variables may work well in the XBlock, and staying with the current format plus new docs may work best for JSinput**

4. Can someone put up a prototype problem in a course? I can't get to this immediately, but could look over something.

**Comment by smarnach** [comment]
@dseaton Thanks for your comments.

1. Fixed points are not introduced with this PR, but have already been a feature before.  They are e.g. used to show the center of mass in the API example in the main directory of this repo.  They could just as well be added to the background image, but apparently someone found it convenient in the past to have this feature.  I reused the `points` list in the configuration for non-fixed points as well, since every other place would probably be confusing.  I had to keep fixed points for backwards compatibility.
2. Yes, the display always shows the properties appropriate for the currently selected object type, and it toggles when switching between vectors and lines.
3. As you have noted yourself, implementing this probably doesn't make sense at the moment, since the expected results are essentially written in JavaScript code, so everyone is free to use variables an reuse them in multiple graders.
4. This would take me a moment, and I currently don't have time to do so either.  I'll see whether we can schedule time for this.

**Comment by ChristopherChudzicki** [comment]
3. This is a bit awkward for points, but you could have one "coords" mechanism:

- coords:[[x0,y0],[x1,y1]] for vectors needs a vector from [x0,y0] to [x1,y1]
- coords:[[x0,y0],[x1,y1]] for segments needs a segment with endpoints at [x0,y0] and [x1,y1], either order
- coords:[[x0,y0],[x1,y1]] for lines (if lines get implemented) needs a line through [x0,y0] and [x1,y1]
- coords:[[x0,y0]] for points needs a point at [x0,y0] 

A little awkard, since coords for points would just be one pair.

**Comment by ChristopherChudzicki** [comment]
@smarnach @dseaton @antoviaque 

I just noticed that `unit_vector_ratio` was removed as a possible setting for lines. The commit message noticed "slope and angle were inconsistent".

`unit_vector_ratio` was intended to solve a problem associated with scales and tick distances:

1. Here tick markers are equally spaced, but represent different lengths on the graph. A line at angle 45 degrees should have slope `1/10`
scale_b
2. Here tick marks are not equally spaced, but represent the same distances on the graph. A line at angle 45 would have slope `2`.
scale_a

If slope is calculated using JSXGraph coordinates as `(y2-y1)/x2-x1)` , that slope will not usually be physically relevant to the background image. `unit_vector_ratio` allowed author to correct for this---basically, telling JSXGraph how to convert between its own native coordinates and the x and y coordinates in the background image.

(The name `unit_vector_ratio` I borrowed from PGFPlots, a plotting package for LaTeX that I used to make most of the background images in my examples).

Was some other method devised for solving this issue?

Here's an example problem where I used `unit_vector_ratio`: action: `https://edge.edx.org/courses/course-v1:CAC+CAC101+FOREVER/jump_to/block-v1:CAC+CAC101+FOREVER+type@vertical+block@3e8bbfbcff3448fbb8d0630aaae07b70`

**Comment by e-kolpakov** [comment]
@dseaton [example problem on edge](https://edge.edx.org/courses/DavidsonNext/Ann101/3T2014/courseware/a51bdc12191c4a5591cdc90f0009fb18/8d5759c04537455eaf63ea663f759c6e/7?activate_block_id=i4x%3A%2F%2FDavidsonNext%2FAnn101%2Fvertical%2F8e05388b77944e05a472fd0744498402)

**Comment by e-kolpakov** [comment]
@ChristopherChudzicki  A couple of sentences that didn't fit well with the rest of the text:

* Unfortunately, I don't have access to the example course you posted.
* Thank you for the line prototype - it was really helpful. The rest of the comment is devoted to problems with slopes, so it might appear as we didn't liked it at all, but that impression (if appeared) would be wrong - we really appreciated your work on it.

The problem with slopes was that it was inconsistent with angle when `unit_vector_ratio` was not equal to 1. The reasons for removing it were:
1. We were unaware of any use cases
2. It was inconsistent with angle (line/vector at angle 45 degrees should have slope=1)
3. In past we had a setting with similar meaning - it was intended to compensate for distortions in the image caused by different JXGraph bounding box and viewport ratio. Later it was removed in favor of automatically calculating bounding box size using viewport ratio. So, one of the reasons for removal was that we thought it is intended to workaround this issue.

So, let me repeat the problem you're solving in my own words, to make sure we're on the same page: the problem is that, if a background image with grid/coordinate axis/other means of measuring distances/angles is used, it might have different scale and X/Y ratio compared to JXGraph. In such a case, it feels more natural to display angle/length/slope relative to background image coordinate grid. `unit_vector_ratio` provided means to compensate for different X/Y ratio between JXGraph grid and background image grid.

If my explanation is correct, there're the following issues with `unit_vector_ratio` as it was implemented by you originally:
1. It affects only slope, but not angle. Slope and angle are essentially a different ways to express the same thing, so they must be consistent: line with 45 deg. angle must have slope of 1, 90 degrees should give slope=infinity and so on. It was not the case - probably you should have updated the "angle" calculation as well. Looking at your initial implementation I can see that you just didn't show angles for lines and slopes for vectors, but we're talking about generic use case now, so it would be best if slopes and angles were made consistent.
2. The problem is slightly more complex, so `unit_vector_ratio` only solves very particular case of having different X/Y ratio, but does not solve different scale. By scale I mean that 1mm vertical real screen distance might be 1 in JXGraph coordinates and 2 in background image coordinates. Same for horizontal distances. Plus they are independent. It might be possible to solve it by setting bounding box manually, but so far it's not possible - we only allow setting "size" of bounding box.

So, if you could fix at least slope-angle inconsistency and send a PR against this branch (or master, if by the time you send it this PR is already merged) we'll be happy to incorporate your changes. Otherwise, we'll discuss with @dseaton to see if we include it in future work.

Hint: An example problem in Notes_and_Examples folder might be very helpful to understand the use case.

**Comment by ChristopherChudzicki** [comment]
@e-kolpakov Thanks for the reply.

Re slope-or-angle for lines, vectors, segments: I agree it makes sense from a coding perspective to allow either slope or angle display for any type. I would be extremely hesitant to deploy an example to students where line that appears to be 45-degrees on screen is labeled as, e.g., 60-degrees in order to be consistent with background image coordinates. 

Re "The problem is slightly more complex ... 1mm vertical real screen distance might be 1 in JXGraph coordinates and 2 in background image coordinates. ... ": I'm not sure I understand. For the purpose of displaying slopes, when is screen distance relevant? We just need to convert from JSXGraph to Background image, right? 
`displayed slope = (jsY2-jsY1)/(jsX2-jsX1) * ( js_to_image_Yscale )/( js_to_image_Xscale )`

I'm excited you all implemented points---hoping to have time to play with that this week.

**Comment by e-kolpakov** [comment]
@ChristopherChudzicki sorry for not responding earlier - yesterday was a little bit packed.

Slope-Angle: as I already said, slope and angle is two ways of expressing the same thing. It is indeed confusing to see the line that appears to have 45 deg. inclination to be labeled as having 60 deg. inclination. It is just as confusing (at least to me and @smarnach who worked on this) to see the line with 45 deg. inclination have slope != 1. The only good solution to this is (likely) to only use background images with X/Y scale == 1 (i.e. square coordinate grid). So, in my opinion, hiding angle and showing slope would not cause a confusion only if a student does not really understand what slope is, which is not what we should aim for.

Problem is more complex - what I was trying to say is that you're only solving one particular problem, but skip other that might emerge. For example, if background image have coordinate grid with step two times smaller than JXGraph one (i.e. 1 unit JXGraph distance == 2 units background image distance) it could be clearly seen that, say, 4cm (bg image coords) segment is reported to have length 2. There are two solutions to the problem: (1) only use bg images without coordinate grid or with coordinate grid with the same scale as JXGraph has or (2) add author-configurable settings to convert between JXGraph and bg image lengths (separate settings for X and Y dimensions).

So, I'm not suggesting you solve the problem as a whole, but there are some outstanding concerns about that feature, and ideally we should aim for a robust solution that addresses both confusion caused by slope-angle distortion and distances distortion.

**Comment by smarnach** [comment]
I merged the current version in this PR.  This isn't meant to cut off the discussion – we can handle all open points in a follow-up.

**Comment by dseaton** [comment]
@e-kolpakov Thank you for the example problem. Question - why am I unable to draw the line? I can use the add button, but cannot initiate the line by drawing.

@ChristopherChudzicki Can you add a problem or two to the "Notes and Examples" folder? Would be great to see the 1:1 image and 1:0.25 images you displayed here in problem format.

@smarnach @e-kolpakov @ChristopherChudzicki We are at a stage of deciding on our primary use case. For the XBlock, I like the idea of focusing on background images only; putting the constraint on authors to match their images to scale/slope. For this iteration on JSinput, how difficult would it be to provide "author-configurable settings" for JXGraph backgrounds?

I am back from travel, and should be responsive this week.

**Comment by e-kolpakov** [comment]
@dseaton 

Drawing lines: I believe @smarnach could answer the question about drawing lines better.

Author configurable settings: about 10 lines of code + testing that it (1) does indeed do what we want and (2) does not break anything else. Not a big deal in human terms :)

**Comment by ChristopherChudzicki** [comment]
@dseaton I'll make a new branch with something similar to the two cases I mentioned. Will try very hard to do it by Wednesday morning. Let me know if sooner is highly desirable.

**Comment by antoviaque** [comment]
@dseaton We can do this, but we are already exceeding the original amount of hours we had budgeted for the improvements (we're at 16h on this now), so that would be eating into the rest of the current tasks time allocation. That's fine with me, but I thought you would like to know before deciding to proceed or not. Let me know - for the next task, I'll either schedule the work on the background discussed here, or the conversion of ActiveTable to an XBlock.

**Comment by dseaton** [comment]
@antoviaque Let's move forward to the next phase - as long as lines are able to be drawn. @smarnach Are we currently able to draw lines?

@ChristopherChudzicki Please keep exploring the issue. My feeling is that this iteration will have to rely primarily on users defining a background image that works with the slope calculations. Let me know if I am misunderstanding the issue.

**Comment by smarnach** [comment]
@dseaton We should be able to draw lines, yes. :)  For lines, I used the code from Chris' prototype, and it worked fine when I tested it.  I will only have time to take a closer look tomorrow morning.

**Comment by dseaton** [comment]
@smarnach Great regarding lines - I was unable to mouse draw lines in the example that @e-kolpakov set up for me. I had been able to in previous examples from Chris.

**Comment by ChristopherChudzicki** [comment]
@dseaton I just created a new branch on my fork, `flexible-slope`: https://github.com/ChristopherChudzicki/jsinput-vectordraw/tree/flexible-slope

- First I updated to be in sync with OpenCraft Master
- I edited two lines in `vectordraw.js` that lets `unit_vector_ratio` alter the slope and angle displayed to user, useful for dealing with different x-axis and y-axis scales.
- I added an example demonstrating this capability (22_BestFit). Here's a link to an example on edge.edx.org: https://edge.edx.org/courses/course-v1:CAC+CAC102+FOREVER/courseware/b65d1993fb884a559c67b7451283a506/48acdb1c6bdd4cce929bfa154f694ec3/2?activate_block_id=block-v1%3ACAC%2BCAC102%2BFOREVER%2Btype%40vertical%2Bblock%40ebe7c6486f274d348bfb8e98ccae0dad 

@dseaton @e-kolpakov Some comments  Regarding the issue of background image x-scale and y-scale vs JSXGraph innate scale:

- `unit_vector_ratio` is good enough for converting slopes and angles, since for this the x-scale conversion and y-scale conversion are not separately important, only their ratio.
- `unit_vector_ratio` is not good enough or converting lengths (One awkward thing about my example is that if `unit_vector_ratio` is used with segments or arrows, the angle is converted correctly but lengths are not). But I'm okay with that ... 
- because anything that requires converting background-image-coordinates / lengths to JSXGraph-coordinates / lengths is **way** too much effort from an author's point of view. `unit_vector_ratio` depends only on the background image and not on JSXGraph at all.

**Comment by antoviaque** [comment]
@dseaton @ChristopherChudzicki Ok, will do.

For the remaining work on this PR, do you want me to still schedule a taks on it next week? That would likely include:

* Reviewing @ChristopherChudzicki 's proposed patch (if we do this, can you send your code in a proper PR? Would you be able to address review comments next week?)
* Checking/updating the example to ensure we can draw lines ( @dseaton to test afterwards to confirm)

If you can confirm if you want this, and whether there is something else that we could help with, I'll schedule it along with the work on ActiveTable. Note that we will need a stable codebase here by the time we start working on converting VectorDrawing to an XBlock - so we should try to resolve all outstanding changes next week, if possible.

**Comment by smarnach** [comment]
@antoviaque I've got a bit of time today, so I can look into the problem with drawing lines.  I know it worked when I tested it locally, but I can see it doesn't work in Eugeny's example, so there is some bug.

**Comment by antoviaque** [comment]
@smarnach Cool, glad to hear : ) In that case yep, definitely worth having a look now.

Thanks!

**Comment by smarnach** [comment]
@dseaton There was indeed a subtle bug caused by JavaScript being a bit special.  I've fixed it (and one more bug I happened to stumble upon) in #24 and uploaded the new version to [the example](https://edge.edx.org/courses/DavidsonNext/Ann101/3T2014/courseware/a51bdc12191c4a5591cdc90f0009fb18/8d5759c04537455eaf63ea663f759c6e/7?activate_block_id=i4x%3A%2F%2FDavidsonNext%2FAnn101%2Fvertical%2F8e05388b77944e05a472fd0744498402).  Please verify again.

**Comment by ChristopherChudzicki** [comment]
@antoviaque Yes, if @dseaton wants I can make a proper PR and can address comments next week. (Will update README first)

**Comment by dseaton** [comment]
Apologies - I thought I responded to this! The new prototype works for me (drawing lines). 

@smarnach The new example works for me, thank you!

@ChristopherChudzicki Upload your latest example for a code review. I believe your suggestion of "hiding length" when unit_vector_ratio!=1 is good for the JSinput version.

@antoviaque @e-kolpakov The XBlock work can start - I view this final feature (unit_vector_ratio) as you all simply code reviewing Chris' logic. We will rely on authors using appropriate background images for the XBlock.

@antoviaque I am late this week, but hopefully we can get things rolling. Did the Active Table XBlock start?

**Comment by antoviaque** [comment]
@dseaton Sounds good. And yes, we did include work on the Active Table XBlock this week - it likely won't be done this sprint, but @smarnach is on it.

**Comment by dseaton** [comment]
@antoviaque Great - keep me posted!

**Comment by antoviaque** [comment]
@dseaton Will do!
